modules/backendworker/backendworker.go

```diff
@@ -0,0 +1,533 @@
+package backendworker
+
+import (
+	"context"
+	"fmt"
+	"hash/fnv"
+	"os"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/gogo/status"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/kv"
+	"github.com/grafana/dskit/ring"
+	"github.com/grafana/dskit/services"
+	backendscheduler_client "github.com/grafana/tempo/modules/backendscheduler/client"
+	"github.com/grafana/tempo/modules/overrides"
+	"github.com/grafana/tempo/modules/storage"
+	"github.com/grafana/tempo/pkg/model"
+	"github.com/grafana/tempo/pkg/tempopb"
+	tempo_util "github.com/grafana/tempo/pkg/util"
+	"github.com/grafana/tempo/pkg/util/log"
+	"github.com/grafana/tempo/tempodb/backend"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc/codes"
+)
+
+const (
+	// ringAutoForgetUnhealthyPeriods is how many consecutive timeout periods an unhealthy instance
+	// in the ring will be automatically removed.
+	ringAutoForgetUnhealthyPeriods = 2
+
+	// We use a safe default instead of exposing to config option to the user
+	// in order to simplify the config.
+	ringNumTokens = 512
+
+	backendWorkerRingKey = "backend-worker"
+
+	reasonCompactorDiscardedSpans = "trace_too_large_to_compact"
+)
+
+var ringOp = ring.NewOp([]ring.InstanceState{ring.ACTIVE}, nil)
+
+type BackendWorker struct {
+	services.Service
+
+	cfg              Config
+	store            storage.Store
+	overrides        overrides.Interface
+	backendScheduler tempopb.BackendSchedulerClient
+
+	workerID string
+
+	// Ring used for sharding tenant index writing.
+	ringLifecycler *ring.BasicLifecycler
+	Ring           *ring.Ring
+
+	subservices        *services.Manager
+	subservicesWatcher *services.FailureWatcher
+}
+
+// var tracer = otel.Tracer("modules/backendworker")
+
+func New(cfg Config, schedulerClientCfg backendscheduler_client.Config, store storage.Store, overrides overrides.Interface, reg prometheus.Registerer) (*BackendWorker, error) {
+	err := ValidateConfig(&cfg)
+	if err != nil {
+		return nil, fmt.Errorf("invalid config: %w", err)
+	}
+
+	w := &BackendWorker{
+		cfg:       cfg,
+		store:     store,
+		overrides: overrides,
+	}
+
+	workerID, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	w.workerID = workerID
+
+	level.Info(log.Logger).Log("msg", "backend worker starting", "worker_id", w.workerID)
+
+	schedulerClient, err := backendscheduler_client.New(cfg.BackendSchedulerAddr, schedulerClientCfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create backend scheduler client: %w", err)
+	}
+	w.backendScheduler = schedulerClient
+
+	if w.isSharded() {
```

Comment:
i really wish you could get rid of this logic. it makes this module seem so much more complicated than it is, but i think we need it as long as we do tenant index building outside the job system?

or can we do a ring of 1 and it just owns all tokens? i can't remember why we did this in the compactor tbh